resources/lemmas.md

```diff
@@ -184,12 +183,19 @@ The rules are applied only when the side-conditions are met.
 These rules are specific to reasoning about EVM programs.
 
 ```k
-    rule (I1 +Int I2) +Int I3 => I1 +Int (I2 +Int I3) when #isConcrete(I2) andBool #isConcrete(I3)
-    rule (I1 +Int I2) -Int I3 => I1 +Int (I2 -Int I3) when #isConcrete(I2) andBool #isConcrete(I3)
-    rule (I1 -Int I2) +Int I3 => I1 -Int (I2 -Int I3) when #isConcrete(I2) andBool #isConcrete(I3)
-    rule (I1 -Int I2) -Int I3 => I1 -Int (I2 +Int I3) when #isConcrete(I2) andBool #isConcrete(I3)
+    //orienting symbolic term to be first, converting -Int to +Int
+    rule I +Int B => B 					+Int I when #isConcrete(I) andBool notBool #isConcrete(B)
+    rule I -Int B => (0 -Int B) +Int I when #isConcrete(I) andBool notBool #isConcrete(B) andBool I =/=Int 0
+    rule A -Int I => A +Int (0 -Int I) when notBool #isConcrete(A) andBool #isConcrete(I)
+    rule 0 -Int (0 -Int X) => X
 
-    rule I1 &Int (I2 &Int I3) => (I1 &Int I2) &Int I3 when #isConcrete(I1) andBool #isConcrete(I2)
+    rule (A +Int I2) +Int I3 => A +Int (I2 +Int I3) when notBool #isConcrete(A) andBool #isConcrete(I2) andBool #isConcrete(I3)
+    rule I1 +Int (B +Int I3) => B +Int (I1 +Int I3) when #isConcrete(I1) andBool notBool #isConcrete(B) andBool #isConcrete(I3)    
+
+    rule I  +Int (0 -Int C) => (0 -Int C) +Int I when 				#isConcrete(I) andBool notBool #isConcrete(C)
+    rule A  +Int (0 -Int C) =>  A -Int C 				 when notBool #isConcrete(A) andBool notBool #isConcrete(C)
+
+    rule I1 &Int (I2 &Int C) => (I1 &Int I2) &Int C when #isConcrete(I1) andBool #isConcrete(I2) andBool notBool #isConcrete(C)
```

Comment:
@denis-bogdanas i cannot see why it is needed to change +Int to -Int and change it back later.

Reply:
It reduces the number of particular cases. Otherwise we would need way more rules to achieve same generality. We would need rules for both (A-I1)+I2, (A+I1)-I2 and many others.

Reply:
to be honest, i'm not sure if the new arithmetic simplification lemmas are equivalent to the existing ones.  also, i'm not sure if the change -Int to +Int and the reverse +Int to -Int will not lead to the infinite loop.  but i hope that the infinite loop will be detected provided that we have no ambiguity of function rule application.

Reply:
I took care of that. Infinite loop is not possible. These are the only 2 rules that rewrite a simpler expression into a more complex one:
```
    rule I -Int B => (0 -Int B) +Int I when #isConcrete(I) andBool notBool #isConcrete(B) andBool I =/=Int 0
    rule A -Int I => A +Int (0 -Int I) when notBool #isConcrete(A) andBool #isConcrete(I)
```
In first rule the side condition `I =/=Int 0` is essential to ensure rule doesn't apply indefinitely.
In 2nd rule a simplification will follow on the RHS. All other rules are crafted in such a way that they cannot match on the result of the 1st rule alone. And you are right we have the option to check determinism now that ensures nothing goes wrong.

Reply:
hmm. @yzhang90 is right.  if the first rule is applied first, then we have `(A + I2) + ((0 - A) + I3)`, which i'm not sure how to be evaluated to `I2 + I3`.

anyway, it seems that we still have the ambiguity, and i think this type of ambiguity cannot be detected by the --deterministic-function feature, since the applicable terms are different. 

@denis-bogdanas just curious, what were the existing ambiguous rules you found initially?

Reply:
@denis-bogdanas my guess is that the ambiguity came from the `infinity` rules initially, which we plan to drop soon, since the rules are not sound.  if that is the case, we can first drop the infinity rules and could minimize the changes to resolve the remaining ambiguity.  what do you think?

Reply:
Ambiguities didn't come only from #infinity. There were many other cases. I didn't keep record of all of them. I would just add new rules for the ambiguous terms that were discovered, until I got too many repetitive rules and I decided to do some low-level normalization instead, like converting -Int to +Int.

The most curious case was indeed related to #infinity. `#infinity -Int #infinity` could rewrite to either #infinity or 0 :D. But because that was part of gas availability logic, it probably didn't have any effect on overall determinism. I fixed this by adding side conditions like
```
    rule N -Int N => 0
      requires #getKLabelString(N) =/=String "#infinity_VERIFICATION"
```

Now to the question why there's no non-determinism in the rule pointed by @yzhang90 : I'm surprised myself. I guess that's because of shady ways kprove rewrite engine works.

If full evaluation is used, then kprove will first evaluate sub-terms then parent terms, and then the `0 -Int A` heating rule will apply and big simplification rule should not be able to apply. But if `substituteAndEvaluate` is used instead, which happens in most cases, only sub-terms affected by substitution are evaluated, and I guess because of this parent term gets evaluated first.

Even if non-determinism arises, it can be fixed by replacing the rule with
```
rule (A +Int I2) +Int ((0 -Int A) +Int I3) => I2 +Int I3
```

I'll activate `--deterministic-functions` for all gnosis projects, my guess is they will all be deterministic.

Anyway, there is one more normalization principle that can be applied and it would fix any possible issues due to rules above: group parentheses always to the right. That is:
```
rule (A+B)+C => A+(B+C) when A is symbolic.
```
This will ensure that symbolic terms accumulate at the left and concrete at the right, where they can add up. So far this was not needed.

I guess you can agree that mathematically lemmas are correct. If you ask me now to get rid of -Int to +Int normalization, I'd have to spend a couple more hours to add other lemmas for parentheses regrouping with +Int and -Int, and to rewrite arithmetic lemmas in verification.k into other forms that will inevitably occur.
So I propose let's not overthink this topic, leave the lemmas as is for now and extend/change only when the need arises. So that we can spend our time on actual gnosis specs. At least these lemmas have an advantage over the previous ones: they are deterministic (modulo kprove engine :D), and they make the specs pass both on our computers and on Jenkins.